plugins/ua_pubsub_realtime_etf.c

```diff
@@ -0,0 +1,259 @@
+        *((__u64 *) CMSG_DATA(controlMsg)) = transmission_time;
+    }
+
+    msgCount = sendmsg(channel->sockfd, &message, 0);
+    if (msgCount < 1) {
+        printf("sendmessage failed");
+        return msgCount;
+    }
+
+    return msgCount;
+}
+
+static int sockErrorQueueProcess(int fd) {
+    /* Function to handle error message queues */
+    uint8_t dataErrorPacket[CMSG_SPACE(sizeof(struct sock_extended_err))];
+    unsigned char errorBuffer[sizeof(txBuffer)];
+    /* Structure for storing the error data */
+    struct cmsghdr* controlErrorMsg;
+
+    /* Structure for gathering of input/output error buffers */
+    struct iovec inputOutputErrorVec = {
+        .iov_base = errorBuffer,
+        .iov_len  = sizeof(errorBuffer)
+    };
+
+    /* Structure for error messages sent and received */
+    struct msghdr messageError = {
+         .msg_iov        = &inputOutputErrorVec,
+         .msg_iovlen     = 1,
+         .msg_control    = dataErrorPacket,
+         .msg_controllen = sizeof(dataErrorPacket)
+    };
+
+    if (recvmsg(fd, &messageError, MSG_ERRQUEUE) == FAILURE_EXIT) {
+        UA_LOG_ERROR(UA_Log_Stdout, UA_LOGCATEGORY_SERVER, "recvmsg failed");
+        return FAILURE_EXIT;
+    }
+
+    controlErrorMsg = CMSG_FIRSTHDR(&messageError);
+    while (controlErrorMsg != NULL) {
+        /* Structure for storing the error queue*/
+        struct sock_extended_err* sockErrorQueue;
+        sockErrorQueue = (struct sock_extended_err *) CMSG_DATA(controlErrorMsg);
+        if (sockErrorQueue->ee_origin == SOCKET_EE_ORIGIN_TRANSMISSION_TIME) {
+            __u64 timeStamp = 0;
+            timeStamp = ((__u64) sockErrorQueue->ee_data << SHIFT_32BITS) + sockErrorQueue->ee_info;
+            switch (sockErrorQueue->ee_code) {
+            case SOCKET_EE_CODE_TRANSMISSION_TIME_INVALID_PARAM:
+                UA_LOG_INFO(UA_Log_Stdout, UA_LOGCATEGORY_SERVER, "packet with timeStamp %llu dropped due to invalid params\n", timeStamp);
+                return 0;
+            case SOCKET_EE_CODE_TRANSMISSION_TIME_MISSED:
+                UA_LOG_INFO(UA_Log_Stdout, UA_LOGCATEGORY_SERVER, "packet with timeStamp %llu dropped due to missed deadline\n", timeStamp);
+                return 0;
+                default:
+                    return -1;
+            }
+        }
+
+        controlErrorMsg = CMSG_NXTHDR(&messageError, controlErrorMsg);
+    }
+
+    return UA_STATUSCODE_GOOD;
+}
+
+
+static void nanoSecondFieldConversion(struct timespec *timeSpecValue) {
+    /* Check if ns field is greater than '1 ns less than 1sec' */
+    while (timeSpecValue->tv_nsec > (SECONDS -1)) {
+        /* Move to next second and remove it from ns field */
+        timeSpecValue->tv_sec  += SECONDS_INCREMENT;
+        timeSpecValue->tv_nsec -= SECONDS;
+    }
+
+}
+
+/* txtime calculation for UDP */
+UA_StatusCode
+txtimecalc_udp(UA_PubSubChannel *channel, UA_ExtensionObject *transportSettings, const UA_ByteString *buf) {
```

Comment:
You have this nice argument with the transportSettings coming for `txtimecalc_udp` and `txtimecalc_ethernet`.
Put the transport settings there. And not into global variables.